### PR TITLE
More challenges

### PR DESCRIPTION
**Pull request by mkuzak**
I've added some more challenges according to #22

### CODE REVIEW

File: 01-rstudio-intro.Rmd [hunk 1]
```diff
@@ -438,18 +438,31 @@ network). R and RStudio have functionality for managing packages:
...
+> .mass
+> MaxLength
+> min-length
+> 2widths
+> celsius2kelvin
+> ```
+
+> ## Challenge 3 {.challenge}
 >
 > Run the code from the previous challenge, and write a command to
```
Comment (tomwright01) [inline 01-rstudio-intro.Rmd]: This needs to be re-phrased "Run the code from challenge 1 ...", alternately, my preference would be to swap new challenge 2 with old challenge 1.

File: 02-project-intro.Rmd [hunk 2]
```diff
@@ -163,6 +163,13 @@ Now we have a good directory structure we will now place/save the data file in t
 >
 > We will load and inspect these data later.
 
+> ## Challenge 2 {.challenge}
+> It is usefull to get some general idea about the dataset, directly from the
```
Comment (tomwright01) [inline 02-project-intro.Rmd]: sp. usefull / useful.

This challenge is only valid if users have shell knowledge beforehand. I think the idea of looking at the data outside of R is a valid one however. Perhaps it could be rephrased "Examine the data without using the R-environment". This would allow use of notepad, excel, explorer etc.

File: 03-seeking-help.Rmd [hunk 3]
```diff
@@ -124,6 +124,14 @@ your issue.
 > What is the difference between the `sep` and `collapse` arguments?
 > 
 
+> ## Challenge 3 {.challenge}
```
Comment (tomwright01) [inline 03-seeking-help.Rmd]: I don't think this lesson needs more challenges. Challenge 4 is ok if the learner thinks to type ```??csv```, I'm not sure if I would think of this, perhaps more guidance (e.g. the ```read.csv()``` function is used to read data from text files...) would be helpful.

File: 04-data-structures-part1.Rmd [hunk 4]
```diff
@@ -339,6 +339,16 @@ In this case, we've explicitly told R that "control" should represented by 1, an
 "case" by 2. This designation can be very important for interpreting the
 results of statistical models!
 
+> ## Challenge 5 {.challenge}
+> Having to factors:
```
Comment (tomwright01) [inline 04-data-structures-part1.Rmd]: This could be more explicit e.g. "The following code creates two variables of type factor:"

File: 04-data-structures-part1.Rmd [hunk 5]
```diff
@@ -339,6 +339,16 @@ In this case, we've explicitly told R that "control" should represented by 1, an
...
+> ```{r, eval=FALSE}
+> sex <- factor(c("M", "F", "F", "M", "M", "F"))
+> age <- factor(sample(100, size=6))
+>```
+> Can you extract values stored by those factors?
```
Comment (tomwright01) [inline 04-data-structures-part1.Rmd]: It is unclear to me if you are requesting the values ```as.numeric(sex)``` or the labelled values ```sex```. Perhaps asking, "what value has been assigned to each label" makes the relationship more explicit.

File: 02-project-intro.Rmd (hunk 2 above)
Comment (mkuzak) [inline 02-project-intro.Rmd]: According to README 'These lesson materials are designed to be run after both the Shell and Git materials'. In this challenge I want people to use their knowledge of shell in new context.

File: 03-seeking-help.Rmd (hunk 3 above)
Comment (mkuzak) [inline 03-seeking-help.Rmd]: Good point, maybe adding a tip about tsv beeing a variant of csv and looking into csv help with help.

Comment (mkuzak) [inline 03-seeking-help.Rmd]: I removed challenge 3 and added more explanation to what used to be challenge 4 before deleting challenge 3.

File: 04-data-structures-part1.Rmd (hunk 5 above)
Comment (mkuzak) [inline 04-data-structures-part1.Rmd]: I reformulated the challenge to make it more clear.

File: 04-data-structures-part1.Rmd (hunk 4 above)
Comment (mkuzak) [inline 04-data-structures-part1.Rmd]: Corrected, more explicit as suggested.

File: 02-project-intro.Rmd (hunk 2 above)
Comment (naupaka) [inline 02-project-intro.Rmd]: Still need to fix the spelling of 'usefull'. I agree with @tomwright01's general sense that more flexibility is better with regard to tool choice, but I also recognize that other parts of this very same lesson discuss creating a `.gitignore` file, for example, which would only happen after a decent amount of time working with shell and git. 

It seems to me that a general principle should be making the lessons as modular as possible but not excessively so.

File: 02-project-intro.Rmd [hunk 6]
```diff
@@ -163,6 +163,13 @@ Now we have a good directory structure we will now place/save the data file in t
...
+> command line, before loading it into R.
+> 1. Check on the command line how large is the file.
```
Comment (naupaka) [inline 02-project-intro.Rmd]: I would reword to "Using the shell, check the size of the file."

File: 02-project-intro.Rmd [hunk 7]
```diff
@@ -163,6 +163,13 @@ Now we have a good directory structure we will now place/save the data file in t
...
+> 2. How many rows of data does it contain?
+> 3. What kind of columns are there?
```
Comment (naupaka) [inline 02-project-intro.Rmd]: Is this asking about the type of the columns (char, numeric, integer, etc)? Or what their names are? Or if they have names? Perhaps you could clarify this in the question since students might be confused.

File: 04-data-structures-part1.Rmd [hunk 8]
```diff
@@ -339,6 +339,17 @@ In this case, we've explicitly told R that "control" should represented by 1, an
 "case" by 2. This designation can be very important for interpreting the
 results of statistical models!
 
+> ## Challenge 5 {.challenge}
+> The following code creates two variables of type factor:
+> ```{r, eval=FALSE}
+> sex <- factor(c("M", "F", "F", "M", "M", "F"))
+> age <- factor(sample(100, size=6))
+>```
+> Try to get character vector of sex values from `sex` factor
```
Comment (naupaka) [inline 04-data-structures-part1.Rmd]: change to: "try to get a character vector of sex values from the `sex` vector (currently of type factor)"

File: 04-data-structures-part1.Rmd [hunk 9]
```diff
@@ -339,6 +339,17 @@ In this case, we've explicitly told R that "control" should represented by 1, an
...
+> and numeric vector from `age`.
+> Does using `as.character()` and `as.numeric()` return results you expect?
```
Comment (naupaka) [inline 04-data-structures-part1.Rmd]: I would reword to:

"Does using the function `as.character()` or the function `as.numeric()` return the results you expect?"

File: 04-data-structures-part1.Rmd [hunk 10]
```diff
@@ -477,8 +488,25 @@ fact many R functions which return complex output store their results in a list.
 > ```
 >
 
+> ## Solution to challenge 5: Factors {.challenge}
+> extracting character vector from factor works as we would expected:
```
Comment (naupaka) [inline 04-data-structures-part1.Rmd]: "extracting character vector from factor" -> "extracting a character vector from a factor vector"

"expected" -> "expect"

File: 04-data-structures-part1.Rmd [hunk 11]
```diff
@@ -477,8 +488,25 @@ fact many R functions which return complex output store their results in a list.
...
+> ```{r, eval=FALSE}
+> as.character(sex)
+> [1] "M" "F" "F" "M" "M" "F"
+> ```
+> but not for factor with numerical values
```
Comment (naupaka) [inline 04-data-structures-part1.Rmd]: "for factor with numerical" -> "for a factor vector with numeric"

**Comment by naupaka** [comment]
Hi @mkuzak I like the new challenge on factors. Factors are a really confusing data type for many learners and I think having an opportunity to test their understanding with this challenge will be useful. I made a few wording suggestions throughout.

File: 02-project-intro.Rmd (hunk 6 above)
Comment (mkuzak) [inline 02-project-intro.Rmd]: Changed.

File: 02-project-intro.Rmd (hunk 7 above)
Comment (mkuzak) [inline 02-project-intro.Rmd]: I thought about data types. I clarified it now.

File: 04-data-structures-part1.Rmd (hunk 8 above)
Comment (mkuzak) [inline 04-data-structures-part1.Rmd]: Agreed, changed.

File: 04-data-structures-part1.Rmd (hunk 9 above)
Comment (mkuzak) [inline 04-data-structures-part1.Rmd]: Changed.

File: 04-data-structures-part1.Rmd (hunk 10 above)
Comment (mkuzak) [inline 04-data-structures-part1.Rmd]: Changed.

File: 03-seeking-help.Rmd [hunk 12]
```diff
@@ -124,6 +124,11 @@ your issue.
...
+> Use help to find out which function and with what parameters should you use
```
Comment (naupaka) [inline 03-seeking-help.Rmd]: Change: "and with what parameters should you use" -> "and with what parameters you could use"

File: 03-seeking-help.Rmd [hunk 13]
```diff
@@ -124,6 +124,11 @@ your issue.
...
+> to load data from csv file with columns delimited with "tab" and decimal point ".".
```
Comment (naupaka) [inline 03-seeking-help.Rmd]: Change: "from csv file" -> "from a csv file"

> columns delimited with "tab" and decimal point "."

Do you mean columns delimited with tab and a decimal point `\t.` or `\t` OR `.`? Perhaps reword to clarify this?

File: 04-data-structures-part1.Rmd [hunk 14]
```diff
@@ -339,6 +339,18 @@ In this case, we've explicitly told R that "control" should represented by 1, an
 "case" by 2. This designation can be very important for interpreting the
 results of statistical models!
 
+> ## Challenge 5 {.challenge}
+> The following code creates two variables of type factor:
+> ```{r, eval=FALSE}
+> sex <- factor(c("M", "F", "F", "M", "M", "F"))
+> age <- factor(sample(100, size=6))
+>```
+> Try to get character vector of sex values from `sex` vector (currently of type
```
Comment (naupaka) [inline 04-data-structures-part1.Rmd]: change: 
"to get character vector" -> "to get a character vector"
"values from `sex` vector" -> "values from the `sex` vector"

File: 04-data-structures-part1.Rmd [hunk 15]
```diff
@@ -339,6 +339,18 @@ In this case, we've explicitly told R that "control" should represented by 1, an
...
+> factor) and numeric vector from `age`.
```
Comment (naupaka) [inline 04-data-structures-part1.Rmd]: change: "and numeric vector" -> "and a numeric vector"

File: 04-data-structures-part1.Rmd [hunk 16]
```diff
@@ -339,6 +339,18 @@ In this case, we've explicitly told R that "control" should represented by 1, an
...
+> Does using function `as.character()` or the function `as.numeric()` return
```
Comment (naupaka) [inline 04-data-structures-part1.Rmd]: change: "Does using function" -> "Does using the function"

File: 03-seeking-help.Rmd (hunk 12 above)
Comment (mkuzak) [inline 03-seeking-help.Rmd]: Are you sure about that? These are the parameters that you need to use if you want the import to work right. I'm not native English speaker though.

File: 04-data-structures-part1.Rmd [hunk 17]
```diff
@@ -339,6 +339,18 @@ In this case, we've explicitly told R that "control" should represented by 1, an
...
+results you expect?
```
Comment (naupaka) [inline 04-data-structures-part1.Rmd]: change: "results you expect" -> "the results you expect"

**Comment by naupaka** [comment]
Looking good @mkuzak! I suggested a couple minor grammar changes, and one point of clarification, after which this should be good to merge.

File: 03-seeking-help.Rmd (hunk 12 above)
Comment (naupaka) [inline 03-seeking-help.Rmd]: Hmmm. I see what you're saying. It just feels a little off though. What about:
"Use help to find a function (and its associated parameters) that you could use to load data from a csv file in..."

Because you could use `read.csv(sep = "\t", ...)` or `read.table(sep = "\t", ...)` and both would work in this case.

File: 03-seeking-help.Rmd [hunk 18]
```diff
@@ -124,6 +124,12 @@ your issue.
 > What is the difference between the `sep` and `collapse` arguments?
 > 
 
+> ## Challenge 3 {.challenge}
+> Use help to find out which function and with what parameters should you use
+> to load data from a csv file in witch columns are delimited with "\t" (tab) and
```
Comment (naupaka) [inline 03-seeking-help.Rmd]: witch -> which

File: 03-seeking-help.Rmd (hunk 12 above)
Comment (tomwright01) [inline 03-seeking-help.Rmd]: I think you should keep the hint here especially as `?read.csv` redirects to `?read.table`

Comment (mkuzak) [inline 03-seeking-help.Rmd]: Agree with both comments. Applied suggestions.

File: 03-seeking-help.Rmd (hunk 18 above)
Comment (mkuzak) [inline 03-seeking-help.Rmd]: Typo fixed.

File: 04-data-structures-part1.Rmd [hunk 19]
```diff
@@ -477,8 +489,25 @@ fact many R functions which return complex output store their results in a list.
 > ```
 >
 
+> ## Solution to challenge 5: Factors {.challenge}
+> extracting character vector from factor vactor works as we would expect:
```
Comment (naupaka) [inline 04-data-structures-part1.Rmd]: "extracting character vector" -> "extracting a character vector"
and
"vector from factor vactor" -> "vector from a factor vector"

File: 03-seeking-help.Rmd [hunk 20]
```diff
@@ -124,6 +124,12 @@ your issue.
 > What is the difference between the `sep` and `collapse` arguments?
 > 
 
+> ## Challenge 3 {.challenge}
+> Use help to find a function (and it's assosiated parameters) that you could use
```
Comment (naupaka) [inline 03-seeking-help.Rmd]: "it's" -> "its"
"assosiated" -> "associated"

File: 03-seeking-help.Rmd [hunk 21]
```diff
@@ -124,6 +124,12 @@ your issue.
...
+> to load data from a csv file in which columns are delimited with "\t" (tab) and
+> decimal point is a "." (period).
```
Comment (naupaka) [inline 03-seeking-help.Rmd]: "decimal point" -> "the decimal point"

I would also add a short bit to the effect of:
"This check for decimal separator is important, especially if you are working with international colleagues, because different countries have different conventions for the decimal point (i.e. comma vs period)."

File: 02-project-intro.Rmd [hunk 22]
```diff
@@ -163,6 +163,15 @@ Now we have a good directory structure we will now place/save the data file in t
 >
 > We will load and inspect these data later.
 
+> ## Challenge 2 {.challenge}
+> It is useful to get some general idea about the dataset, directly from the
+> command line, before loading it into R. Understanding better the dataset
```
Comment (naupaka) [inline 02-project-intro.Rmd]: "Understanding better the dataset" -> "Understanding the dataset better"

File: 02-project-intro.Rmd [hunk 23]
```diff
@@ -163,6 +163,15 @@ Now we have a good directory structure we will now place/save the data file in t
...
+> will come handy when making decisions on how to load it in R. Use Unix shell
```
Comment (naupaka) [inline 02-project-intro.Rmd]: "Use Unix shell" -> "Use the command-line shell"

**Comment by naupaka** [comment]
Just a few more minor fixes.

**Comment by naupaka** [comment]
One letter more and everything else looks good to merge

**Comment by naupaka** [comment]
Thanks @mkuzak!
